targets/TARGET_Freescale/TARGET_KLXX/TARGET_KL46Z/flash_api.c

```diff
@@ -0,0 +1,225 @@
+    0xd301428b, 0x1ac0008b, 0x08434152, 0xd301428b, 0x1ac0004b, 0x1a414152, 0x4601d200, 0x46104152,
+    0xe05d4770, 0xd0000fca, 0x10034249, 0x4240d300, 0x22004053, 0x0903469c, 0xd32d428b, 0x428b0a03,
+    0x22fcd312, 0xba120189, 0x428b0a03, 0x0189d30c, 0x428b1192, 0x0189d308, 0x428b1192, 0x0189d304,
+    0x1192d03a, 0x0989e000, 0x428b09c3, 0x01cbd301, 0x41521ac0, 0x428b0983, 0x018bd301, 0x41521ac0,
+    0x428b0943, 0x014bd301, 0x41521ac0, 0x428b0903, 0x010bd301, 0x41521ac0, 0x428b08c3, 0x00cbd301,
+    0x41521ac0, 0x428b0883, 0x008bd301, 0x41521ac0, 0x0843d2d9, 0xd301428b, 0x1ac0004b, 0x1a414152,
+    0x4601d200, 0x41524663, 0x4610105b, 0x4240d301, 0xd5002b00, 0x47704249, 0x105b4663, 0x4240d300,
+    0x2000b501, 0x46c046c0, 0xb430bd02, 0x1e644674, 0x1c647825, 0xd20042ab, 0x5d63461d, 0x18e3005b,
+    0x4718bc30, 0xfffffffe, 0xffffffff, 0xfffffffe, 0x00000000, 0x00000000, 0x00000000, 0x00000000,
+    0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000
+};
+
+static const flash_algo_t flash_algo_config = {
+    .init = 0x1,
+    .uninit = 0x45,
+    .erase_sector = 0x9d,
+    .program_page = 0xb5,
+    .static_base = 0x5f0,
+    .algo_blob = FLASH_ALGO,
+};
+
+static const sector_info_t sectors_info[] = {
+    {0x0, 0x400},
+};
+
+static const flash_target_config_t flash_target_config = {
+    .page_size  = 0x100,
+    .flash_start = 0x0,
+    .flash_size = 0x40000,
+    .sectors = sectors_info,
+};
+
+/*
+    This binary blob (thumb code) sets r9 (static base) as the code we are jumping to
+    is PIC (position independent code).
+
+    These are the instructions (r0 is a pointer to arg_t):
+        push {r5, lr, r4}
+        mov r5, r9
+        push {r5}
+        ldr r5, [r0, #20]
+        ldr r3, [r0, #16]
+        mov r9, r3
+        ldr r3, [r0, #12]
+        ldr r2, [r0, #8]
+        ldr r1, [r0, #4]
+        ldr r0, [r0, #0]
+        blx r5
+        pop {r5}
+        mov r9, r5
+        pop {r4-r5, pc}
+        bx r14
+*/
+static uint32_t jump_to_flash_algo[] = {
+    0x464DB530,
+    0x6945B420,
+    0x46996903,
+    0x688268C3,
+    0x68006841,
+    0xBC2047A8,
+    0xBD3046A9
+};
+
+static uint32_t get_sector_index(flash_t *obj, uint32_t address)
+{
+    // check where address belongs to
+    size_t sector_index = 0;
+    size_t sectors_count = sizeof(obj->target_config->sectors) / sizeof(sector_info_t);
+    for (; sector_index < sectors_count; sector_index++) {
+        if ((address > obj->target_config->sectors[sector_index].start) && 
+            (address < (obj->target_config->sectors[sector_index].start +obj->target_config->sectors[sector_index].size))) {
+            break;
+        }
+    }
+    return sector_index;
+}
+
+int32_t flash_init(flash_t *obj)
+{
+    core_util_critical_section_enter();
+
+    obj->flash_algo = &flash_algo_config;
+    obj->target_config = &flash_target_config;
+
+    args_t arguments = {
```

Comment:
This has extra parameters that might need to be used. Also this might need to be called once before each operation.
https://www.keil.com/pack/doc/CMSIS/Pack/html/algorithmFunc.html